test/pretty-printer/ppTest.ml

```diff
@@ -0,0 +1,85 @@
+open Base
+open Mtt.Ast
+open Mtt.PrettyPrinter
+
+(* QCheck generator for the arbitrary (and most likely invalid) expressions *)
+let generator =
+  QCheck.Gen.(
+    sized
+    @@ fix (fun self size ->
+           let lowercase_id =
+             string_size ~gen:(char_range 'a' 'z') (return 1)
+           in
+           let regular_id id = Mtt.Id.R.mk id in
+           let modal_id id = Mtt.Id.M.mk (id ^ "'") in
+           match size with
+           | 0 ->
+               oneof
+                 [
+                   return Expr.Unit;
+                   (* won't work until expressions with free variables can be pretty-printed *)
+                   (* map
+                      (fun s -> Expr.VarL (Mtt.Id.R.mk s))
+                      (string_size ~gen:(char_range 'a' 'z') (return 1)); *)
+                   map (fun idg -> Expr.VarG (modal_id idg)) lowercase_id;
+                 ]
+           | size ->
+               let open Expr in
+               let unary_node f = map f (self (size - 1)) in
+               let binary_node f = map2 f (self (size / 2)) (self (size / 2)) in
+               oneof
+                 [
+                   binary_node pair;
+                   unary_node fst;
+                   unary_node snd;
+                   binary_node app;
+                   map3 func
+                     (map regular_id lowercase_id)
+                     (return Type.Unit)
+                     (self (size - 1));
+                   unary_node box;
+                   map3 letc
+                     (map regular_id lowercase_id)
+                     (self (size / 2))
+                     (self (size / 2));
+                   map3 letbox
+                     (map modal_id lowercase_id)
+                     (self (size / 2))
+                     (self (size / 2));
+                 ]))
+
+let arbitrary_ast =
+  let print_ast t = [%sexp_of: Expr.t] t |> Sexp.to_string_hum in
+  let shrink_ast =
+    let open QCheck.Iter in
+    function
+    | Expr.Unit | Expr.VarL _ | Expr.VarG _ -> empty
+    | Expr.Fst pe | Expr.Snd pe -> return pe
+    | Expr.Pair (e1, e2) -> of_list [ e1; e2 ]
+    | Expr.Fun (_, _, body) -> return body
+    | Expr.App (fe, arge) -> of_list [ fe; arge ]
+    | Expr.Box e -> return e
+    | Expr.Let (_, bound_e, body) -> of_list [ bound_e; body ]
+    | Expr.Letbox (_, boxed_e, body) -> of_list [ boxed_e; body ]
```

Comment:
Looks like `shrink_ast` should be recursive, otherwise it can only shrink by forgetting information, not changing it. Sometimes we might want to keep the head constructor, but shrink the components.

One thing to think about -- we don't have to keep the constructor intact all the time; but that's a bit more involved, so if there is no immediate solution, then let's just scratch this part.